contracts/StreamManager.vy

```diff
@@ -1,259 +1,464 @@
 
-    prefunded_stream_life: uint256 = max(MIN_STREAM_LIFE, block.timestamp - start_time)
-    assert max_stream_life >= prefunded_stream_life  # dev: prefunded stream life too large
-    assert funded_amount >= prefunded_stream_life * amount_per_second  # dev: not enough funds
+@external
+def create_stream(
+    token: IERC20,
+    amount: uint256,
+    products: DynArray[bytes32, MAX_PRODUCTS],
+    min_stream_life: uint256 = MIN_STREAM_LIFE,
+) -> uint256:
+    """
+    @dev Create a streaming payment to `controller` using the given `token`, pre-funded with
+        a given `amount` of `token`, in exchange for the provisiong of `products` to the caller,
+        for at least `min_stream_life` length of time.
+    @notice This function starts the lifecycle of the Stream datastructure. It cannot be revoked
+        once called until at least `MIN_STREAM_LIFE` has passed. The stream's parameters flow
+        through a series of `validators` to check for their validity and compute the cost of
+        provisiong the `products` in the given `token`. The `token` must be accepted by this
+        contract.
+    @param token An ERC20-compatible token that this contract allows to create streams for.
+    @param amount The amount of `token` that should be pre-funded for this stream.
+    @param products An array of the product codes this stream should pay for. The product codes are
+        treated as application-specific parameters and have no special treatment by this contract.
+        Typically, validators are employted to do the specific processing necessary to compute the
+        stream rate for the newly created stream.
+    @param min_stream_life A safety parameter designed to ensure that the computed stream rate does
+        not exceed the value of `amount / min_stream_life` tokens per second. Defaults to
+        `MIN_STREAM_LIFE` and is validated not to be below that amount, which is the minimum length
+        *any* new stream can be created for (based on the time it takes provision the `products`).
+    @return stream_id The globally unique identifier for the newly created stream.
+    """
+    assert min_stream_life >= MIN_STREAM_LIFE  # dev: stream life not long enough
+    assert self.token_is_accepted[token]  # dev: token not accepted
 
     assert extcall token.transferFrom(  # dev: transfer fail
-        msg.sender, self, funded_amount, default_return_value=True
+        msg.sender, self, amount, default_return_value=True
     )
 
-    stream_id: uint256 = self.num_streams[msg.sender]
-    self.streams[msg.sender][stream_id] = Stream({
+    # Check all validators for any unacceptable or incorrect stream parameters
+    amount_per_second: uint256 = 0
+    for validator: Validator in self.validators:
+        # NOTE: Validator either raises or returns a funding rate to add to the total
+        amount_per_second += extcall validator.validate(msg.sender, token, amount, products)
```

Comment:
I think of the set of validators as the way to check and validate stream parameters to compute the subtotal cost of a stream. There may be different product codes handled by different validators

How do you think I can make this clearer?

Reply:
Each validators gets the full list of products, maybe that is why I am confused. If it were the case of like mapping products to validators, it might be less confusing

Reply:
I kind of agree validators could be done a bit better, maybe. And they're conceptually a little confusing. Just *adding* to the price doesn't seem very useful.

When I first thought of validators, I was thinking about things that can approve/reject/modify the terms of the stream. For instance, it might make sense to write one up that sets discounts for certain users/times. And you might have another you might want to check an exchange rate with uniswap before returning a price.

For these concepts to work, you'd need to be able to modify the price, not just add to it.

Reply:
Actually, this gave me some ideas that I think can eliminate a few footguns and overall make it simpler. Will refactor and push it up

Reply:
implemented in [9674830](https://github.com/ApeWorX/ApePay/pull/112/commits/9674830d7044aeecf499a1dd10946f2dfdcedb32) and further commits, please review

Reply:
I'm not sure how this changes much on the validator front.  Now instead of validators only adding to the stream cost, they can only add to the stream duration?

That's fine with me, if that's what you're going for. I'm just not really sure how this makes validators more useful or understandable.